tests/cpu_bugs/xen_pv.pm

```diff
@@ -0,0 +1,394 @@
+        },
+        "prctl,ibpb" => {
+            default => {
+                expected => {'cat /proc/cmdline' => ['spectre_v2_user=prctl,ibpb']},
+                unexpected => {'cat /proc/cmdline' => ['spectre_v2_user=on', 'spectre_v2_user=off', 'spectre_v2_user=prctl ', 'spectre_v2_user=seccomp', 'spectre_v2_user=seccomp,ibpb', 'spectre_v2_user=auto']}
+            }
+        },
+        seccomp => {
+            default => {
+                expected => {'cat /proc/cmdline' => ['spectre_v2_user=seccomp']},
+                unexpected => {'cat /proc/cmdline' => ['spectre_v2_user=on', 'spectre_v2_user=off', 'spectre_v2_user=prctl', 'spectre_v2_user=prctl,ibpb', 'spectre_v2_user=seccomp,ibpb', 'spectre_v2_user=auto']}
+            }
+        },
+        "seccomp,ibpb" => {
+            default => {
+                expected => {'cat /proc/cmdline' => ['spectre_v2_user=seccomp,ibpb']},
+                unexpected => {'cat /proc/cmdline' => ['spectre_v2_user=on', 'spectre_v2_user=off', 'spectre_v2_user=prctl', 'spectre_v2_user=prctl,ibpb', 'spectre_v2_user=seccomp ', 'spectre_v2_user=auto']}
+            }
+        },
+        auto => {
+            default => {
+                expected => {'cat /proc/cmdline' => ['spectre_v2_user=auto']},
+                unexpected => {'cat /proc/cmdline' => ['spectre_v2_user=on', 'spectre_v2_user=off', 'spectre_v2_user=prctl', 'spectre_v2_user=prctl,ibpb', 'spectre_v2_user=seccomp', 'spectre_v2_user=seccomp,ibpb']}
+            }
+        },
+    },
+};
+
+sub do_check {
+    my $secnario = shift;
+    my $foo      = $secnario->{default};
+    if ($foo->{expected}) {
+        while (my ($cmd, $lines) = each %{$foo->{expected}}) {
+            foreach my $expected_string (@{$lines}) {
+                if ($expected_string ne "") {
+                    #my $ret = script_run("$cmd | grep \"$expected_string\"");
+                    my $ret = ssh_vm_cmd("$cmd | grep \"$expected_string\"");
+                    record_info("ERROR", "Can't found a expected string.", result => 'fail') unless $ret eq 0;
+                } else {
+                    print "This expection is empty string, skip";
+                }
+
+            }
+        }
+    }
+    if ($foo->{unexpected}) {
+        while (my ($cmd, $lines) = each %{$foo->{unexpected}}) {
+            foreach my $unexpected_string (@{$lines}) {
+                if ($unexpected_string ne "") {
+                    #my $ret = script_run("$cmd | grep \"$unexpected_string\"");
+                    my $ret = ssh_vm_cmd("$cmd | grep \"$unexpected_string\"");
+                    record_info("ERROR", "found a unexpected string.", result => 'fail') unless $ret ne 0;
+                } else {
+                    #Debug what output be report.
+                    assert_script_run("xl dmesg | grep -A 10 \"Speculative\"");
+                    print "This unexpection is empty string, skip";
+                }
+
+            }
+        }
+    }
+}
+
+sub do_test {
```

Comment:
not sure why we need this function , it is only thing what you call inside `run` and you doing it once . So why not just move all this code into `run` and have one function less ? Also not sure why you passing `mitigation_list` as input param, variable is visible in scope of `do_test` so just refer directly to it